### DISCUSSION TITLE
getChildren() is broken

### DISCUSSION THREAD

**Discussion by kara-hyyyper** [pimcore/pimcore#17725] [Q&A / Support]
When I call getChildren() to get all the child variants of a parent, it doesn't iterate over every child for some unknown reason.
To force iteration I have to do this every time I call getChildren()
```
 foreach ($parentObject->getChildren([DataObject::OBJECT_TYPE_VARIANT, DataObject::OBJECT_TYPE_OBJECT, AbstractObject::OBJECT_TYPE_OBJECT, AbstractObject::OBJECT_TYPE_VARIANT]) as $childObject) {}
```
This is getting kind of crazy.

Why are my objects different types of objects to begin with when I created them in the exact same way? Some are DataObjects, some are AbstractObjects, with absolutely no pattern to it. And either way, why doesn't getChildren just iterate over every data type by default?

**Comment by muratbinerbay** [comment]
I don't understand the second part your question so I can't really help there.

But for the first part:
Are you using an empty `foreach` loop to load the children? If you are not immediately iterating over `getChildren()` then you need to ass `->load()` so the collection is loaded. If you do that the empty loop should not be needed

Mentioned in the upgrade notes:

> Calling getChildren/getSiblings on AbstractObject, Document and Asset now returns unloaded listing. If the list is not traveresed immediately, then it is required to call load() explicitily.

https://pimcore.com/docs/platform/Pimcore/Installation_and_Upgrade/Upgrade_Notes/#dataobjectsdocuments-

**Comment by podarcis** [comment]
There might be a misconception here: The `getChildren()` returns the direct children for given types. In older Pimcore versions, this was *excluding* variants, in newer Pimcore versions, it's all types.

And you don't have to specify the types twice, as they are effectively the same. The real place of them is `AbstractObject`, so just do:
```php
foreach ($parentObject->getChildren([AbstractObject::OBJECT_TYPE_OBJECT, AbstractObject::OBJECT_TYPE_VARIANT]) as $childObject) {
  // do something
}
```

The class hierarchy internally is (usually) `YourClass` -> `Concrete` -> `DataObject` -> `AbstractObject`. So your data objects of one class are all of the same type.

**Reply by kara-hyyyper** [reply]
From what I've noticed, the getChildren() function seems to work fine when working with calculators. But it is very buggy when I'm running my script with the getChildren() function. Your solution worked for me one time. I updated some information in my objects with the data importer tool. After that the getChildren did not return all the children yet again, when using my script. Not sure what's going on here. It definitely has something to do with it being in a script, since its working fine in the calculators. Thanks for your reply none-the-less.
